### DISCUSSION TITLE
just my humble opinion.

### DISCUSSION THREAD

**Discussion by vipsland** [juicedata/juicefs#2657] [General]
may i add that (imho) u dont need to spend so much time with other metadata stuff like forestdb (super edge case).

tdlr; what i am suggesting is focusing on making this for extreme large scale solution will be where juicefs stand out and not really trying to do so many connectors and expending resources when the problem scenario can be truly improved just focusing on the main  issue, solving large scale storage cheaply, reliably, easy to maintain cost effectively and with peace of mind production. officially support tikv and minio.

for me, performance, reliability, security shld be the core focus of data storage layer
suggestion : 
a) https://github.com/cloudwego/netpoll <- is juicefs using this kind of network layer? just curious. can be faster this way. and of course io_uring stuff (in future)

b) OFFICIAL supported metadata should be tikv
c) OFFICIAL supported storage shld be minio

let's be honest here. just the implementation of juicefs is overkill if it's not for extreme large scale data storage. meaning, there are easier ways to maintain and operate other storage solutions (just get a bigger NAS) than focus on juicefs if not for extremely large scale cost saving cloud deployment.

both tikv and minio are designed for extreme cases and they have long term maintenance. for all other storages, imho, it's... relatively more cost efficient NOT to deploy them. postgres / mysql for reliability? rationally, (imho) at certain scalability threshold, just using postgres / mysql is more profitable than deploying juicefs + mysql etc in terms of maintenance effort too.

from what i can see / tell, the larger u go, the maintenance effort on all other metadata storage solutions are going to be 
a) not optimum for speed or cost of maintenance
b) not practical. (redis / codis-pika for faster processing of metadata storage? at what kind of storage scale?)

dedup pls. thx :D

**Comment by davies** [comment]
Thanks for sharing with us.

Currently, we support three kinds of meta engine (Redis, SQL database and transactional KV), they provide different capability and fit in different situations. Most of people don't have peta bytes of data, they may find the value of JuiceFS very well.

TiKV is an officially supported for large scale, and we know companies use that to manage peta bytes of data and billions of files. We will continue work with community, also the TiKV community, to make sure TiKV and JuiceFS can work well for large scale workload.

To my knowledge, Minio may not be the best solution for large scale, other cloud services or Ceph RADOS work even better in large scale. People can pick what's fit well in their situation, this kind of freedom is great.
